sharutils-4.9/doc/uuencode.1
'.\" Copyright (c) 1980, 1990 The Regents of the University of California.
'.\" All rights reserved.
'.\"
'.\" Redistribution and use in source and binary forms, with or without
'.\" modification, are permitted provided that the following conditions
'.\" are met:
'.\" 1. Redistributions of source code must retain the above copyright
'.\"    notice, this list of conditions and the following disclaimer.
'.\" 2. Redistributions in binary form must reproduce the above copyright
'.\"    notice, this list of conditions and the following disclaimer in the
'.\"    documentation and/or other materials provided with the distribution.
'.\" 3. All advertising materials mentioning features or use of this software
'.\"    must display the following acknowledgement:
'.\"	This product includes software developed by the University of
'.\"	California, Berkeley and its contributors.
'.\" 4. Neither the name of the University nor the names of its contributors
'.\"    may be used to endorse or promote products derived from this software
'.\"    without specific prior written permission.
'.\"
'.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
'.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
'.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
'.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
'.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
'.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
'.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
'.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
'.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
'.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
'.\" SUCH DAMAGE.
'.\"
'.\" Modified from
'.\"     @(#)uuencode.1	6.9 (Berkeley) 4/23/91
'.\"
.TH uuencode 1
.SH NAME
uuencode, uudecode \- encode a binary file, or decode its representation
.SH SYNOPSIS
.B uuencode
[\-m] [ file ] name
.PP
.B uudecode
[-o outfile] [ file ]...
.SH DESCRIPTION
.I Uuencode
and
.I uudecode
are used to transmit binary files over channels that support only
simple ASCII data.
.PP
.I Uuencode
reads
.I file
(or by default the standard input) and writes an encoded version to
the standard output, using only printable ASCII characters.
The encoded output begins with a header, for use by
.IR uudecode ,
which records the mode of the input file and suggests
.I name
for the decoded file that will be created.  (If
.I name
is
.I /dev/stdout
then
.I uudecode
will decode to standard output.)  The encoding has the format
documented at uuencode(5), unless the option
.I \-m
is given, when
.B base64
encoding is used instead.
.PP
.B Note:
.I uuencode
uses buffered input and assumes that it is not hand typed from a tty.
The consequence is that at a tty, you may need to hit Ctl-D several times
to terminate input.
.PP
.I Uudecode
transforms uuencoded
.I files
(or standard input) into the original form.
The resulting file is named
.I name
(or
.I outfile
if the \-o option is given)
and will have the mode of the original file except that setuid
and execute bits are not retained.  If
.I outfile
or
.I name
is /dev/stdout the result will be written to standard output.
.I Uudecode
ignores any leading and trailing lines.  The program determines from the
header which of the two supported encoding schemes was used.
.SH EXAMPLES
The following example packages up a source tree, compresses it,
uuencodes it and mails it to a user on another system.
When
.I uudecode
is run on the target system, the file ``src_tree.tar.Z'' will be
created which may then be uncompressed and extracted into the original
tree.
.PP
.IP
.NF
tar cf \- src_tree \&| compress \&|
uuencode src_tree.tar.Z \&| mail sys1!sys2!user
.FI
.LP
.SH SEE ALSO
compress(1), mail(1), uucp(1), uuencode(5)
.SH STANDARDS
This implementation is compliant with P1003.2b/D11.
.SH BUGS
If more than one file is given to
.I uudecode
and the \-o option is given or more than one
.I name
in the encoded files are the same the result is probably not what is expected.
.PP
The encoded form of the file is expanded by 37% for UU encoding and by 35%
for base64 encoding (3 bytes become 4 plus control information).
.SH REPORTING BUGS
Report bugs to <bug-gnu-utils@gnu.org>.  Please put
.I sharutils
or
.I uuencode
in the subject line.  It helps to spot the message.
.SH HISTORY
The
.I uuencode
command appeared in BSD 4.0.


noflushd/NEWS
* noflushd 2.7.5 (09. May 2005)

  - Fix handling of raid and similar meta devices on 2.6 kernels.

* noflushd 2.7.4 (08. Aug 2004)

  - Fix lousy spindown times on 2.6 kernels.
  - Fix SCSI support on 2.6 kernels.
  - Fix pid detection in generic startup script.
  - Document quirks with hotplugged drives.

* noflushd 2.7.3 (09. Apr 2004)

  - Fix race condition that might leave kupdated/pdflushd disabled
    aftern noflushd has terminated.
  - Fix spindown when all disks are auto-detected.

* noflushd 2.7.2 (16. Mar 2004)

  - A change in 2.7.1 erroneously flushed the buffer cache before
    spindown, causing frequent spinups. This change has been reverted.
    
* noflushd 2.7.1 (07. Mar 2004)

  - Fix immediate spinup on 2.6 kernels.
  - Elevate SCSI support from compile-time to command line option.
  - Bug fixes in the disk statistics parser for 2.6 kernels.
  - Ship kernel patch to track down which processes cause spinups.

* noflushd 2.7 (20. Jan 2004)

  - Support Linux kernel 2.6.x.
  - Default to monitorring all disks unless specific disks are given
    on the command line. This change should yield a sane default
    behaviour for a majority of users without requiring explicit
    configuration.

* noflushd 2.6.3 (13. May 2002)

  - Bugfix for arm and other archs with weird /proc/partitions layout.

* noflushd 2.6.2 (20. Apr 2002)

  - Add IDE raid support.
  - Add timeout logging script.

* noflushd 2.6.1 (03. Feb. 2002)

  - Fix BUG when using noflushd with disks on second IDE channel on 2.4
    kernels.

* noflushd 2.6 (22. Jan 2002)

  - Document noflushd's suckage with any journaling fs, including ext3.
  - New feature prevents spindown if machine is in interactive use.
  - Add hack fix some '0 spindown time' uglies.

* noflushd 2.5 (06. Aug 2001)

  - Make sure hotplugged disks get synced properly.
  - Only sync present partitions. (Fixes high CPU usage bug on devfs
    systems.)
  - Generic startup skript works on latest 2.4 kernels.

* noflushd 2.4 (13. Mar 2001)

  - Timeout lists are back to a working state.
  - Support very latest 2.4 kernels.

* noflushd 2.3 (12. Jan 2001)

  - Support latest 2.4 kernels. (Leaves kernels 2.3.99pre8 to 2.4.0-test8
    unsupported.)
  - Fix multi-disk support with long bdflush timeouts.

* noflushd 2.2 (22. Dec 2000)

  - Fixed bug in spindown code. Now actually spins down. Duh!

* noflushd 2.1 (20. Dec 2000)

  - Disks get synced prior to spindown.
  - Improved devfs support.

* noflushd 2.0 (14. Nov 2000)

  - noflushd now comes with its own syncer that takes over the task from
    kupdate when the first read-write mounted disk is spun down. This
    enhances spindown times in a multidisk environment at the expense of
    less smooth and more bursty background flushing when the noflushd 
    syncer is in use.

  - Conflicting entries in 2.2-style /proc/stat (e.g. /dev/sda and /dev/hda)
    won't cause noflushd to give up anymore. Instead, noflushd will treat
    both disks pretty much like one, i.e. they're either both up or spun
    down, no matter if activity only goes to one of them.

  - Compilation environment switched to automake/autoconf.

  - When installing via 'make install', noflushd is no longer added to
    any runlevel. 

  - noflushd moved to SourceForge. The project homepage is located at
    http://noflushd.sourceforge.net, a mailing list is available via
    noflushd-devel@lists.sourceforge.net. Visit the homepage for 
    subscription information.



sharutils-4.9/src/uuencode.c

static const char cright_years_z[] =

/* uuencode utility.
   Copyright (C) */ "1994, 1995, 1996, 2002, 2005, 2007, 2010";

/* Free Software Foundation, Inc.

   This product is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3, or (at your option)
   any later version.

   This product is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, see:  http://www.gnu.org/licenses.
*/

/* Copyright (c) 1983 Regents of the University of California.
   All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions
   are met:
   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
   3. All advertising materials mentioning features or use of this software
      must display the following acknowledgement:
	 This product includes software developed by the University of
	 California, Berkeley and its contributors.
   4. Neither the name of the University nor the names of its contributors
      may be used to endorse or promote products derived from this software
      without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
   ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
   IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
   ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
   FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
   DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
   OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
   HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
   LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
   OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
   SUCH DAMAGE.  */

/* Reworked to GNU style by Ian Lance Taylor, ian@airs.com, August 93.  */

/*=======================================================\
| uuencode [INPUT] OUTPUT				 |
| 							 |
| Encode a file so it can be mailed to a remote system.	 |
\=======================================================*/

#include "local.h"

#if HAVE_LOCALE_H
#else
# define setlocale(Category, Locale)
#endif
#define _(str) gettext (str)

#if __CYGWIN__
# include <fcntl.h>
# include <io.h>
#endif

#define	RW (S_IRUSR | S_IWUSR | S_IRGRP | S_IWGRP | S_IROTH | S_IWOTH)

static struct option longopts[] =
{
  { "base64", 0, 0, 'm' },
  { "version", 0, 0, 'v' },
  { "help", 0, 0, 'h' },
  { NULL, 0, 0, 0 }
};

static inline void try_putchar __P ((int));
static void encode __P ((void));
static void usage __P ((int))
#if defined __GNUC__ && ((__GNUC__ == 2 && __GNUC_MINOR__ >= 5) || __GNUC__ > 2)
     __attribute__ ((noreturn))
#endif
;

/* The name this program was run with. */
const char *program_name;

/* Pointer to the translation table we currently use.  */
const char *trans_ptr;

/* The two currently defined translation tables.  The first is the
   standard uuencoding, the second is base64 encoding.  */
const char uu_std[64] =
{
  '`', '!', '"', '#', '$', '%', '&', '\'',
  '(', ')', '*', '+', ',', '-', '.', '/',
  '0', '1', '2', '3', '4', '5', '6', '7',
  '8', '9', ':', ';', '<', '=', '>', '?',
  '@', 'A', 'B', 'C', 'D', 'E', 'F', 'G',
  'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O',
  'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W',
  'X', 'Y', 'Z', '[', '\\', ']', '^', '_'
};

const char uu_base64[64] =
{
  'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H',
  'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P',
  'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X',
  'Y', 'Z', 'a', 'b', 'c', 'd', 'e', 'f',
  'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n',
  'o', 'p', 'q', 'r', 's', 't', 'u', 'v',
  'w', 'x', 'y', 'z', '0', '1', '2', '3',
  '4', '5', '6', '7', '8', '9', '+', '/'
};

/* ENC is the basic 1 character encoding function to make a char printing.  */
#define ENC(Char) (trans_ptr[(Char) & 077])

static inline void
try_putchar (c)
         int c;
{
  if (putchar (c) == EOF)
    error (EXIT_FAILURE, 0, _("Write error"));
}

/*------------------------------------------------.
| Copy from IN to OUT, encoding as you go along.  |
`------------------------------------------------*/

static void
encode ()
{
  register int n;
  int finishing = 0;
  register char *p;
  char buf[46];  /* 45 should be enough, but one never knows... */

  while ( !finishing && (n = fread (buf, 1, 45, stdin)) > 0 )
    {
      if (n < 45)
        {
          if (feof (stdin))
            finishing = 1;
          else
            error (EXIT_FAILURE, 0, _("Read error"));
        }

      if (trans_ptr == uu_std)
        putchar (ENC (n));

      for (p = buf; n > 2; n -= 3, p += 3)
        {
          try_putchar (ENC (*p >> 2));
          try_putchar (ENC (((*p << 4) & 060) | ((p[1] >> 4) & 017)));
          try_putchar (ENC (((p[1] << 2) & 074) | ((p[2] >> 6) & 03)));
          try_putchar (ENC (p[2] & 077));
        }

      if (n > 0)  /* encode the last one or two chars */
        {
          char tail = trans_ptr == uu_std ? ENC ('\0') : '=';

          if (n == 1)
            p[1] = '\0';

          try_putchar (ENC (*p >> 2));
          try_putchar (ENC (((*p << 4) & 060) | ((p[1] >> 4) & 017)));
          try_putchar (n == 1 ? tail : ENC ((p[1] << 2) & 074));
          try_putchar (tail);
        }

      try_putchar ('\n');
    }

  if (ferror (stdin))
    error (EXIT_FAILURE, 0, _("Read error"));
  if (fclose (stdin) != 0)
    error (EXIT_FAILURE, errno, _("Read error"));

  if (trans_ptr == uu_std)
    {
      try_putchar (ENC ('\0'));
      try_putchar ('\n');
    }
}

static void
usage (status)
     int status;
{
  if (status != 0)
    fprintf (stderr, _("Try `%s --help' for more information.\n"),
	     program_name);
  else
    {
      printf (_("Usage: %s [INFILE] REMOTEFILE\n"), program_name);
      fputs (_("\n\
  -m, --base64    use base64 encoding as of RFC1521\n\
      --help      display this help and exit\n\
      --version   output version information and exit\n"), stdout);
      /* TRANSLATORS: add the contact address for your translation team! */
      printf (_("Report bugs to <%s>.\n"), PACKAGE_BUGREPORT);
    }
  exit (status);
}

int
main (argc, argv)
     int argc;
     char *const *argv;
{
  int opt;
  struct stat sb;
  int mode;

  /* Set global variables.  */
  trans_ptr = uu_std;		/* Standard encoding is old uu format.  */

  program_name = argv[0];
  setlocale (LC_ALL, "");

  /* Set the text message domain.  */
  bindtextdomain (PACKAGE, LOCALEDIR);
  textdomain (PACKAGE);

  while (opt = getopt_long (argc, argv, "hm", longopts, (int *) NULL),
	 opt != EOF)
    {
      switch (opt)
	{
	case 'h':
	  usage (EXIT_SUCCESS);

	case 'm':
	  trans_ptr = uu_base64;
	  break;

	case 'v':
	  printf ("%s (GNU %s) %s\n", basename (program_name),
		  PACKAGE, VERSION);
	  /* xgettext: no-wrap */
	  printf (_("Copyright (C) %s Free Software Foundation, Inc.\n\
This is free software; see the source for copying conditions.  There is NO\n\
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.\n"),
		  cright_years_z);
	  exit (EXIT_SUCCESS);

	case 0:
	  break;

	default:
	  usage (EXIT_FAILURE);
	}
    }

  switch (argc - optind)
    {
    case 2:
      /* Optional first argument is input file.  */
      {
#if S_IRWXU != 0700
        choke me - Must translate mode argument
#endif

	FILE *fp = freopen (argv[optind], FOPEN_READ_BINARY, stdin);
	if (fp != stdin)
	  error (EXIT_FAILURE, errno, _("fopen-ing %s"), argv[optind]);
	if (fstat (fileno (stdin), &sb) != 0)
	  error (EXIT_FAILURE, errno, _("fstat-ing %s"), argv[optind]);
	mode = sb.st_mode & (S_IRWXU | S_IRWXG | S_IRWXO);
	optind++;
	break;
      }

    case 1:
#if __CYGWIN__
      if (! isatty (STDIN_FILENO))
       setmode (STDIN_FILENO, O_BINARY);
#endif

      mode = RW & ~umask (RW);
      break;

    case 0:
    default:
      usage (EXIT_FAILURE);
    }

  if (printf ("begin%s %o %s\n", trans_ptr == uu_std ? "" : "-base64",
	      mode, argv[optind]) < 0)
    error (EXIT_FAILURE, errno, _("Write error"));

  encode ();

  if (ferror (stdout) ||
      printf (trans_ptr == uu_std ? "end\n" : "====\n") < 0 ||
      fclose (stdout) != 0)
    error (EXIT_FAILURE, errno, _("Write error"));

  exit (EXIT_SUCCESS);
}
